daffodil-runtime1/src/main/scala/org/apache/daffodil/infoset/DataValue.scala

```diff
@@ -0,0 +1,209 @@
+/*
+ * Licensed to the Apache Software Foundation (ASF) under one or more
+ * contributor license agreements.  See the NOTICE file distributed with
+ * this work for additional information regarding copyright ownership.
+ * The ASF licenses this file to You under the Apache License, Version 2.0
+ * (the "License"); you may not use this file except in compliance with
+ * the License.  You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package org.apache.daffodil.infoset
+
+import java.lang.{ Boolean => JBoolean, Number => JNumber, Long => JLong, Double => JDouble, String => JString, Float => JFloat, Byte => JByte, Integer => JInt, Short => JShort }
+import java.math.{ BigDecimal => JBigDecimal, BigInteger => JBigInt }
+import org.apache.daffodil.calendar.DFDLCalendar
+import scala.language.implicitConversions
+import org.apache.daffodil.calendar.DFDLDateTime
+import org.apache.daffodil.calendar.DFDLDate
+import org.apache.daffodil.calendar.DFDLTime
+import org.apache.daffodil.util.Maybe
+import org.apache.daffodil.processors.UseNilForDefault
+import org.apache.daffodil.exceptions.Assert
+import org.apache.daffodil.util.Maybe.One
+import org.apache.daffodil.util.Maybe.Nope
+import java.io.ObjectOutputStream
+import java.io.ObjectInputStream
+import java.net.URI
+
+/*
+ * These traits are used for the phantom type X. When considering the type of a
+ * DataValue. See the comment on the DataValue class below.
+ */
+trait NonNullable extends Nullable;
```

Comment:
Ok, I am confused. 

Nullable has to do with the DataValue having a "Null" that is the NoValue sentinel. I suggest name is misleading. Null suggests Nil which is not what this is about at all. It's about the fact that the DataValue class represents a hybrid of a Maybe[T] type with a constructor system that prevents constructing (or selecting out) anything but one of our approved values. So to me DataValuePrimitive and NoValue and DataValuePrimitiveOrNoValue are the right names. "Null" just confuses the heck out me for some reason. I believe it is going to work to have NoValue be both an object name and type name. 

tUseNilForDefault is about the sentinel hack we use to indicate that if defaulting, we should instead set the nilled flag. Suggest UseNilForDefaultSentinelType.

Please add scaladoc to this effect so that these things don't get conflated in readers minds.

Reply:
UseNilForDefault is not a NoValueType.

Almost everywhere we have DataValuePrimitiveOrNoValue, UseNilForDefault is not a legal value.

The `trait NonNullable extends Nullable` also turns out to be important in order to get auto upcasting to work. Once you get your mind around this (which I have yet to do), adding a third trait to the hierarchy is simple.

This contravariance stuff is confusing enough as is. I would prefer to keep it standard inheritance hierarchy that we flip upside down; instead of trying to combine it with the singleton definitions.

Reply:
UseNilForDefault is not a NoValueType.

Almost everywhere we have DataValuePrimitiveOrNoValue, UseNilForDefault is not a legal value.

The `trait NonNullable extends Nullable` also turns out to be important in order to get auto upcasting to work. Once you get your mind around this (which I have yet to do), adding a third trait to the hierarchy is simple.

This contravariance stuff is confusing enough as is. I would prefer to keep it standard inheritance hierarchy that we flip upside down; instead of trying to combine it with the singleton definitions.